glusterd2/events/eventhandler.go

```diff
@@ -0,0 +1,129 @@
+package events
+
+import (
+	"sort"
+	"strings"
+	"sync"
+)
+
+// Handler is a function that is registered to be called when an event happens
+// event is the name of the event that happened and data is any extra data that
+// was attached to the event as a json document
+type Handler func(*Event)
+
+// HandlerID is returned when a Handler is registered. It can be used to unregister a registered Handler.
+type HandlerID uint64
+
+var (
+	handlers struct {
+		wg sync.WaitGroup
+
+		sync.RWMutex
+		chans map[HandlerID]chan<- *Event
+		next  HandlerID
+	}
+)
+
+func init() {
+	handlers.chans = make(map[HandlerID]chan<- *Event)
+}
+
+func addHandler(c chan<- *Event) HandlerID {
+	handlers.Lock()
+	defer handlers.Unlock()
+
+	id := handlers.next
+	handlers.chans[id] = c
+	handlers.next++
+
+	return id
+}
+
+func delHandler(id HandlerID) chan<- *Event {
+	handlers.Lock()
+	defer handlers.Unlock()
+
+	c, ok := handlers.chans[id]
+	if !ok {
+		return nil
+	}
+	delete(handlers.chans, id)
+	return c
+}
+
+// Register a Handler to be called when the given events happen.
+// If no events are specified, handler function is called for all events
+// Handlers need to be thread safe, as they can be called concurrently when
+// multiple events arrive at the same time.
+func Register(h Handler, events ...string) HandlerID {
+	in := make(chan *Event)
+	id := addHandler(in)
+
+	go func() {
+		handlers.wg.Add(1)
+		handleEvents(in, h, events...)
```

Comment:
A general comment here on when go routines are spawned will help. One goroutine is spawned per handler and inside that goroutine, a new gorotuine is spawned per event.